File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int,int> nge;
        stack<int> st;

        for(int i=0;i<nums2.size();i++){
            int num = nums2[i];
            while(!st.empty() && st.top() < num) 
            {
                nge[st.top()] = num;
                st.pop();
            }
            st.push(num);
        }

        while(!st.empty()) 
        {
            nge[st.top()] = -1;
            st.pop();
        }

        vector<int> ans(nums1.size());
        for(int i = 0; i < nums1.size(); i++) {
            ans[i] = nge[nums1[i]];
        }
        return ans;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int [] a= new int[10001];

        Arrays.fill(a,-1);
        Stack<Integer> st= new Stack <>();
        for(int i=0;i<nums2.length;i++){
            while(!st.isEmpty() && st.peek()< nums2[i]){
                a[st.pop()]= nums2[i];
            }
            st.push(nums2[i]);
        }
        int [] a1 = new int[nums1.length];
        for(int i=0;i<nums1.length;i++){
            a1[i] = a[nums1[i]];
        }
        return a1;
    }
}